src/crimson/osd/pg.cc

```diff
@@ -940,33 +940,43 @@ seastar::future<> PG::submit_error_log(
     log_entries, t, peering_state.get_pg_trim_to(),
     peering_state.get_min_last_complete_ondisk());
 
-    set<pg_shard_t> waiting_on;
-    for (auto &i : get_acting_recovery_backfill()) {
+  return seastar::do_with(log_entries, t, set<pg_shard_t>{},
+    [this, rep_tid](auto& log_entries, auto& t,auto& waiting_on) mutable {
+    return seastar::do_for_each(get_acting_recovery_backfill(),
+      [this, log_entries, t=std::move(t), waiting_on, rep_tid]
+      (auto& i) mutable {
       pg_shard_t peer(i);
-      if (peer == pg_whoami) continue;
+      if (peer == pg_whoami) {
+        return seastar::now();
+      }
       ceph_assert(peering_state.get_peer_missing().count(peer));
       ceph_assert(peering_state.has_peer_info(peer));
       auto log_m = crimson::make_message<MOSDPGUpdateLogMissing>(
-                   log_entries,
-                   spg_t(peering_state.get_info().pgid.pgid, i.shard),
-                   pg_whoami.shard,
-                   get_osdmap_epoch(),
-                   get_last_peering_reset(),
-                   rep_tid,
-                   peering_state.get_pg_trim_to(),
-                   peering_state.get_min_last_complete_ondisk());
-      send_cluster_message(peer.osd, std::move(log_m), get_osdmap_epoch());
+        log_entries,
+        spg_t(peering_state.get_info().pgid.pgid, i.shard),
+        pg_whoami.shard,
+        get_osdmap_epoch(),
+        get_last_peering_reset(),
+        rep_tid,
+        peering_state.get_pg_trim_to(),
+        peering_state.get_min_last_complete_ondisk());
       waiting_on.insert(peer);
-    }
-    waiting_on.insert(pg_whoami);
-    log_entry_update_waiting_on.insert(
-      std::make_pair(rep_tid, log_update_t{std::move(waiting_on)}));
-    return shard_services.get_store().do_transaction(
-      get_collection_ref(), std::move(t))
-      .then([this] {
+      return shard_services.send_to_osd(peer.osd,
+                                        std::move(log_m),
+                                        get_osdmap_epoch());
```

Comment:
`send_to_osd()` is now chained inside `CommonPGPipeline::process` which is an `OrderedExclusivePhase`. IIUC this should be able to serialize message sending across phases (i.e. https://github.com/ceph/ceph/pull/52192#issuecomment-1620961130 case [A]).

So, is it guaranteed that different tids must be disjoint to enter `CommonPGPipeline::process`?

The correct serialization can be verified by adding another log (as finish-send-osd) at L968 below after `send_to_osd()` to see if sends across tids are really serialized, so that:
* start-send-osd rep_tid=72057594037928058
* finish-send-osd rep_tid=72057594037928058
* start-send-osd rep_tid=72057594037928059
* finish-send-osd rep_tid=72057594037928059

If the above order is correct when the reordering happens, at least we can be sure that https://github.com/ceph/ceph/pull/52192#issuecomment-1620961130 is unrelated to this issue as [A] is already enforced.

Reply:
> `send_to_osd()` is now chained inside `CommonPGPipeline::process` which is an `OrderedExclusivePhase`. IIUC this should be able to serialize message sending across phases (i.e. [#52192 (comment)](https://github.com/ceph/ceph/pull/52192#issuecomment-1620961130) case [A]).
> 
> So, is it guaranteed that different tids must be disjoint to enter `CommonPGPipeline::process`?

This should be guaranteed. However, it seems like this is not true since we have racing senders.

> The correct serialization can be verified by adding another log (as finish-send-osd) at L968 below after `send_to_osd()` to see if sends across tids are really serialized, so that:
> ...
> If the above order is correct when the reordering happens, at least we can be sure that [#52192 (comment)](https://github.com/ceph/ceph/pull/52192#issuecomment-1620961130) is unrelated to this issue as [A] is already enforced.


After adding (start/finish-send) it can be seen that the re-ordering is happening in the sender side. I will verify that  [A] is actually enforced.
See: https://gist.github.com/Matan-B/b28ae9fe2ce43efe58f8ce4555c28ead?permalink_comment_id=4623285#gistcomment-4623285